### ISSUE TITLE
Advanced Node and Express Challenges

### ISSUE THREAD

**Issue by QuincyLarson** [freeCodeCamp/CurriculumExpansion#49]
@Em-Ant is in charge of coordinating the expansion of these challenges, but he needs your help.

We need to create a list of topics we'd like to cover. I suspect we should cover streams and advanced file system methods. Are there any Node.js enthusiasts who would like to contribute to this?

Current challenge list (these need to be converted to proper challenge names and several of these topics will require more than one challenge):
- [ ] cookies
- [ ] sessions
- [ ] make API calls in the server with the http module (or better with request)
- [ ] express.router and structure the app with modules
- [ ] the fs module to read/write files
- [ ] some basic stream/pipe concepts to handle files transmission ❓
- [ ] view engines ❓
- [ ] Authorization with passport

**Comment by alayek** [comment]
Might we want to cover RxJS? It's an advanced topic, and requires some background in Reactive Programming, and goes really well with the understanding of stream. Heavily use functional programming.

But this is a portable skill. Other languages like .NET, Python, Java, Swift, Go, Android Java etc. have their Rx<Language> library.

A lot of back-end jobs appreciate knowledge in Rx concepts (most popular would be Netflix's RxJava library and Microsoft's Rx for .NET).

**Comment by Em-Ant** [comment]
@Greenheart and me suggested to have this section to cover the arguments that cannot be handled in the express section (#44 section 2) for time reasons, but that can be useful (necessary) for the final projects. Maybe **Advanced Express** would be a better name . We were thinking to cover :

- probably move cookies here 
- sessions 
- make API calls in the server with the `http` module (or better with `request`)
- `express.router` and structure the app with modules
- the `fs` module to read/write files
- some basic stream/pipe concepts to handle files transmission :question: 
- view engines :question: 
- Authorization with `passport`

@alayek @QuincyLarson what do you think ? @alayek Could RxJs fit into this ( I don't know it, but I'd like to learn...) ?

**Comment by QuincyLarson** [comment]
@Em-Ant OK - these can be called "Advanced Express Challenges" then. 

@BerkeleyTrue what are your thoughts on teaching RxJS as part of the advanced Express challenges?

**Comment by BerkeleyTrue** [comment]
RxJS should not be part of the curriculum.

**Comment by QuincyLarson** [comment]
@alayek based on @BerkeleyTrue's opinion (he has worked extensively with RxJS in the past) and the fact that we're running behind on this, I suggest we move forward with the topics @Em-Ant suggested. I've updated my OP to feature these as a checklist.

**Comment by AngelKyriako** [comment]
@Em-Ant Let me know regarding the work needed to be done on the scopes listed above.

Adding JWT authentication supplementary to passport might be a good idea as well, since it is helpful for making a universal device agnostic authorization module.

Finally, If it is not out of scope, we could include sails.js(node.js version of ruby on rails) on the final parts of these challenges. It removes a lot of headaches (and introduces a few extras) which really helps in kickstarting very quick a backend app

**Comment by Em-Ant** [comment]
@AngelKyriako Thanks for offering your help. I really need it.
I haven't started working on this and probably I would not be able to it them before Sept 16 without help.

We need to create these challenges on HyperDev. It should be something like [the basic express challenges](https://hyperdev.com/#!/project/west-horse). The exercises should be remotely testable. Ideally you should create them having in mind how to check the results. The project I linked has a background API which is contained in a npm module. If you want I can give you the source code. I cannot publish it  because Hyperdev seems  be not working at the moment. [This project](https://hyperdev.com/#!/project/venom-deer) (the mongoose challenges) has not been moved to a npm package yet, so you can see how it works. We use [this prototype](https://hyperdev.com/#!/project/marsh-glazer) to implement a draft version of the tests. In the final version, the testing should happen on a specific FCC page.

In [this document](https://hackmd.io/EwE2EMGYDNwDgLQCMCmBjA7AgLANl8AgJzogICsGUAjOCCAAyQpxA===?view) there are the descriptions of the basic express challenges. We should aim to have about 10, max 15 challenges.

If you want you can start building them. I can help with Hyperdev if necessary. You can reach me on gitter, or we can schedule an hangout. 
The list above is only a reference. You can add (or even remove) things if you want. Keep in mind that these exercises should  be preparatory for the FCC backend projects, and should be solvable in 2-5 min. If you have troubles with the testing system you can focus on the challenges. We can work on the remote testing later.

About sails.js, it sounds interesting, but probably it would require its own section. Let's see what @QuincyLarson thinks about it.

Let me know if you are going to work on this. Thank you very much.

**Comment by AngelKyriako** [comment]
@QuincyLarson It should be fine, the only problem is that I have limited time, so it might not be fast.

As I can understand from the prototype the testing occurs client-side.

@Em-Ant Lets arrange for a call or chat session at some point so that I can be on the same page on how this thing is gonna be implemented.

How can I find you on gitter ? Should I create a room for the advanced node.js challenges ?

**Comment by atjonathan** [comment]
@AngelKyriako https://gitter.im/QuincyLarson :stuck_out_tongue_winking_eye:

**Comment by AngelKyriako** [comment]
Thanks, but I was referring to emanuele

**Comment by atjonathan** [comment]
@AngelKyriako :joy: 
In that case, https://gitter.im/em-ant

**Comment by Em-Ant** [comment]
@AngelKyriako Thank you. 
You can find a description of how these challenges should work in the **Action Plan** and **Challenges Structure** paragraphs of #44. 
See you tomorrow.

**Comment by QuincyLarson** [comment]
Hi @AngelKyriako, could you give us a quick update on your progress?

**Comment by AngelKyriako** [comment]
@QuincyLarson  I liked how the flow of the cat app challenge was on the css challenges and I was thinking of doing something similar. I am building an basic chat app with the challenge list in mind (express, passport, views, request .etc). However, I am not focusing on  structuring the app in modules right now. Making the code straightforward is going to help challengers to understand it easier. Plus I am not sure if separating the app into directories is even an option on the browser editor.

Once it is done with @Em-Ant  we will take our scissors and setup the flow of the challenges that will result into a chat app in the end.

Let me know what you think

**Comment by Em-Ant** [comment]
@AngelKyriako souns good to me :+1: . Are you working on Hyperdev ? If yes, can I see the WIP project (I'm just curious...).

**Comment by AngelKyriako** [comment]
@Em-Ant I am working locally, I could send you a zip in the gitter chat in the afternoon

**Comment by QuincyLarson** [comment]
@AngelKyriako Just a quick update - we are getting very close to being able to launch all of these new back end challenges.

@Em-Ant tells me that you have made significant progress on these. I am excited to see them.

We would like to get these challenges onto beta.freecodecamp.com by Wednesday. Do you have time to make a final push of effort to finish these up? @Em-Ant and I can handle the data entry component if you can get @Em-Ant everything he needs.

Thank you again for your hard work on this. I'm very excited about these advanced topics, and I think our community will learn a lot from them.

**Comment by AngelKyriako** [comment]
@QuincyLarson I had a call with @Em-Ant where he explained me in more details the testing process.

I have created a chat project with everything listed above
whats left:
1) pick the parts to be stripped of for the challenges
2) unit test those parts & create client side test strings
3) include internal files into an npm module so that it is not editable on hyperdev

I will let you know when I have something you can check.

**Comment by QuincyLarson** [comment]
@AngelKyriako OK - that sounds good. I'm looking forward to seeing your challenges :)

**Comment by AngelKyriako** [comment]
this is the project that is made bit by bit with the challenges:
https://hyperdev.com/#!/project/luminous-dart

this is the test framework where I am writing the client side unit tests
https://hyperdev.com/#!/project/summer-bite

@QuincyLarson you could checkout the first project for a first review of the challenges, Each server#.js contains the corresponding challenge with the solution (ctrl-f '#challenge') and has a detailed description and guidelines on the top comments.

@Em-Ant I have started with the tests, but I found it troublesome to port the code into strings. Is there any feature on the fcc client where I can evaluate a javascript string during initialization ? You can checkout the TestManager object on the bottom of main.js on the second project which would idealy could be evaluated into the fcc client side. That way the unit test list can be tweaked quite easily.

As far as the challenges go I have two major concerns:
1) I do not think creating a challenge to introduce a global error handling mechanic is possible. Any ideas on how this could be tested? My opinion on this end is that it could be easily skipped since it is quite based on dev preferences anyway.

2) hyperdev does not support file uploads, For that reason the only filesystem related challenge (challenge 12) I came up with is to create a route that will pipe a default avatar already saved on the server side(.eg the internal fcc-npm module). Any thoughs on that ?

Please, let me know before I continue with the more time consuming unit tests.

**Comment by Em-Ant** [comment]
@AngelKyriako I'm afraid you cannot use Your *TestManager* . The ajax call must happen into the test string. Look at the tests here https://hyperdev.com/#!/project/marsh-glazer.
You have to use that format, i.e. the string to be *eval*uated should return a function that executes the ajax call, make assertions on the results, and return a promise or use a node callback. If you need help we can arrange a call tonight. Let me know.

**Comment by QuincyLarson** [comment]
@AngelKyriako thank you for your patience with my slow response. As @Em-Ant said, we want the tests to be in version control in Free Code Camp's repository, rather than on the Hyperdev project itself. Here's an example of what one of these looks like:

```
    {
      "id": "9d764678c441eaafaeb5bdef",
      "title": "Chain Middleware to Create a Time Server",
      "description": [
        "Middleware can be mounted at a specific route using app.METHOD('<path>', mware). Middleware can also be chained inside route definition.",
        "Look at the following example :",
        "<code>app.get('/user', function(req, res, next){</code>",
        "<code>  req.user = getTheUserSync();  // Hypotetical Synchronous operation</code>",
        "<code>  next();</code>",
        "<code>}, function(req, res) {</code>",
        "<code>  res.send(req.user)</code>",
        "<code>})</code>",
        "This approach is useful to split the server operations into smaller units. That leads a to a better app structure, and the possibility to reuse code in different places. This approach can be also used to perform some validation on the data. At each point of the middleware stack you can block the execution of the current chain and pass control to functions specifically desinged to handle errors. Or you can pass control to the next matching route, to handle special cases. We will see how in the advanced express section.",
        "In the route app.get('/now', ...) chain a middleware function and the final handler. In the middleware function you should add the current time to the request object in the req.time key. You can use new Date().toString(). In the handler, respond with a json object, taking the structure {time: req.time}.",
        "Hint: The test will not pass if you don’t chain the middleware. If you mount the function somewhere else, the test will fail, even if the output result is correct."
      ],
      "challengeSeed": [],
      "tests": [
        {
          "text" : "The /now endpoint doesn't have mounted middleware",
          "testString" : "() => ($.get('https://west-horse.hyperdev.space' + '/_api/chain-middleware-time').then(function(data){ assert.equal(data.stackLength, 2); }).promise())"
        },
        {
          "text" : "The /now endpoint doesn't return a time that is +/- 20 secs from now",
          "testString" : "() => ($.get('https://west-horse.hyperdev.space' + '/_api/chain-middleware-time').then(function(data){ var now = new Date(); assert.isAtMost(Math.abs(new Date(data.time) - now), 20000); }).promise())"
        }
      ],
      "solutions": [],
      "hints": [],
      "type": "waypoint",
      "challengeType": 0,
      "translations": {
        "de": {
          "description": [],
          "title": ""
        },
        "fr": {
          "description": [],
          "title": ""
        },
        "pt-br": {
          "description": [],
          "title": ""
        },
        "ru": {
          "description": [],
          "title": ""
        }
      }
    }
```

@Em-Ant and I can import your challenge directions and tests ourselves if you can structure them in a way that follows a similar pattern:

```     
   {
          "text" : "The /now endpoint doesn't have mounted middleware",
          "testString" : "() => ($.get('https://west-horse.hyperdev.space' + '/_api/chain-middleware-time').then(function(data){ assert.equal(data.stackLength, 2); }).promise())"
        }
```

By the way, the "https://west-horse.hyperdev.space" URL is temporary, and will be dynamically replaced by the URL for the camper's Hyperdev project.

@Em-Ant actually wrote these tests himself, and can answer any questions you may have about them. 

Thank you for your help. The descriptions you've written in your projects are excellent, and I will go ahead and start importing these myself. 

The main work left is for you to translate your TestManager tests into a format that can be run from Free Code Camp and hit the right endpoints on the camper's HyperDev instance.

**Comment by AngelKyriako** [comment]
@QuincyLarson I will continue with the unit tests with the test manager and I will port them to strings later on, Hopefully, I can find an automated way to achieve that. This is not a problem at all.

What I want both your opinions is these two concerns:

1) I do not think creating a challenge to introduce a global error handling mechanic is possible. Any ideas on how this could be tested? My opinion on this end is that it could be easily skipped since it is quite based on dev preferences anyway.

2) hyperdev does not support file uploads, For that reason the only filesystem related challenge (challenge 12) I came up with is to create a route that will pipe a default avatar already saved on the server side(.eg the internal fcc-npm module). Any thoughts on that ? How about adding 1-3 challenges with file upload via an external file service like cloudinary or something ? Would this be too much or the default avatar pipe wll suffice ?

**Comment by Em-Ant** [comment]
@AngelKyriako 
1) what do you mean for global error handling ? If you think it is very hard/impossible to do for me you can skip it. Your project is very 'rich' already.
2) According to [this](https://hyperdev.com/help/faqs/) you can write to `/tmp` even if what you upload will be destroyed in 15 mins. Don't know if it's good enough for what you want to do.

**Comment by AngelKyriako** [comment]
1) I mean a error handling middleware for prettyfying errors. I have seen error handling listed in the challenges list in the documentation somewhere, but since its not on the list on top of this ussue I guess its cool.

2) This is a great find, Will take a look on what to do with it after the tests are done.

**Comment by QuincyLarson** [comment]
@AngelKyriako I agree with @Em-Ant on skipping global error handling for now. We can always add this later if we think of a way to implement this.

**Comment by AngelKyriako** [comment]
Hello guys,

 I am in the process on the unit test development, However, I am sticking with my test manager for now since the tests are a bit complicated. We can use this http://www.freeformatter.com/javascript-escape.html later on to escape them.

They should be ready soon enough, except of the final challenge with the default avatar pipe. This challenge will require us to setup an npm module which will be added into the package.json. This module will include the  internal modules which will not be accessible by the challenger & the default avatar file.

@Em-Ant checkout my fork of the test app: https://hyperdev.com/#!/project/summer-bite
You may open the console so that u can see in the logs that the tests have passed. The way this app is setup, it does not support asynchronous assertions. That means that, if an assert fails in the response callback of a request, it will not be seen into the UI. Will this problem exist in the production free code camp as well ?

@QuincyLarson I will need access to a repository for building the internal module. We do not even have to publish it to npm for now. Just include it via git url (https://docs.npmjs.com/files/package.json#git-urls-as-dependencies)

Should we have an online call to communicate the integration with production or is there a specific QA procedure that fcc follows?

Let me know,

Ciao

**Comment by Em-Ant** [comment]
@AngelKyriako I don't know if I understand exactly what you mean, but using promises the assertion is synchronous : it happens when the promise resolves.

**Comment by AngelKyriako** [comment]
Lets move this to gitter so that I can explain you in detail.

**Comment by AngelKyriako** [comment]
@QuincyLarson I had a call with @Em-Ant and we agreed on waiting for the feature/curriculum-expansion branch to be merged back to staging for now.

Once, its setup I can start with importing the tests into the fcc codebase

**Comment by QuincyLarson** [comment]
@AngelKyriako @Em-Ant I don't understand. We won't be able to merge feature/curriculum-expansion into staging until feature/curriculum-expansion is stable. I believe @BerkeleyTrue plans to deploy staging to production soon, while our curriculum expansion effort will take longer to finish and QA.

**Comment by QuincyLarson** [comment]
@AngelKyriako @Em-Ant I've imported the advanced express challenges into our seed files, so you can start the process of adding tests to them once they're ready. They're located in: /seed/challenges/06-information-security-and-quality-assurance

**Comment by Em-Ant** [comment]
There is a misunderstanding ( my fault, sorry )... 

@AngelKyriako I thought we agreed on waiting for the **backend view PR** ( https://github.com/FreeCodeCamp/FreeCodeCamp/pull/11058 ) to be merged into staging to **QA your tests locally**. We have a couple of bugs, but it's almost done and I think it will be released soon.

The branch `feature/curriculum-expansion` **must have all tests in place  before being released**.

@AngelKyriako there is a modification proposed by Berkeley to the test format. I'm pushing it now on the branch `feature/curriculum-expansion`. Please re-pull it in your local repository, so you have the latest test version.

**Comment by AngelKyriako** [comment]
@QuincyLarson  I guess I misunderstood what branch I was waiting for. @Em-Ant I will do so, since hyperdev is back online now.

**Comment by QuincyLarson** [comment]
I'm keeping this issue open for now since @Greenheart is working on QA'ing these challenges and potentially finishing them this weekend.

**Comment by QuincyLarson** [comment]
@AngelKyriako @Greenheart I haven't heard from either of you on this so I'm closing this issue and we can handle these on an challenge-by-challenge basis over on staging.

**Comment by AngelKyriako** [comment]
@QuincyLarson I understood that the rest of the team will take care of integrating the challenges into the FCC repo. Do you need more help/pointers from me on that ?

**Comment by QuincyLarson** [comment]
@AngelKyriako Actually, we haven't made any progress on this since you stopped working on it. We have managed to build the interactive back end projects, though. We are currently QA'ing the entire back end curriculum.

I know you are a busy guy, but would you be interested in picking up where you left off and finishing these advanced Node/Express challenges?

**Comment by AngelKyriako** [comment]
@QuincyLarson I need an additional hook(before) in the json that describes what unit tests to run for the challenge, so that it runs a javascript string before all the test cases.
.eg https://github.com/AngelKyriako/FccAdvancedNodejsExpressApp/blob/master/testApp/public/seeds.js

This will let us update the unit tests with less pain when the challenges need polishing.

I could help on adding this improvement on the backend testing framework, but it feels that it will be quicker if one of the creators did it.

I will also need a call or a readme to guide me on how to run the fcc locally to test that the challenges work. Last time I checked that was not that easy as the framework was still under development.

**Comment by QuincyLarson** [comment]
@AngelKyriako are all these challenges working properly on GoMix.com (formerly HyperDev.com)? If so, what is the current URL for both the boilerplate and the tester? The tester we're using for all the other projects only has one test on it: https://narrow-plane.gomix.me/

I can talk with @BerkeleyTrue about how we might go about adding such a Git Hook, and whether this is the best way to go about this.

**Comment by AngelKyriako** [comment]
@QuincyLarson since the tests are client side I did not upload them on GoMix, The challenges on the other hand were developed on hyperdev in a first place, to always have in mind its restrictions: https://gomix.com/#!/project/luminous-dart. Notice that it does not run since its on the challenge 1 state.

You can copy the content of these files: https://github.com/AngelKyriako/FccAdvancedNodejsExpressApp/tree/master/app/solutions in server.js or index.ejs have it run on any challenge# state u want.

I do not know about this tester. But the previous tester was not working properly for me(had problems with the callbacks on the testing functions) so I made this: https://github.com/AngelKyriako/FccAdvancedNodejsExpressApp/tree/master/testApp simple test app to test it locally myself.

I don't get the meaning of 'Git Hook'.
All I need is a before hook in the test JSON blueprint that will run before all cases.
Something like what most frameworks do with: before beforeEach after afterEach hooks.

**Comment by QuincyLarson** [comment]
@AngelKyriako thank you for the additional detail. 

I will talk with @Em-Ant about how we can add these to the current tester.

"Git hook" was a typo - I meant to type "hook."

**Comment by Em-Ant** [comment]
@QuincyLarson  I'm looking at @AngelKyriako 's code. If I understood correctly, he needs a script to run before each test. It's probably possible to add this functionality to our testing framework on gomix, but more important this also needs to be implemented in the FCC backend code (by Berkeley).

**Comment by BerkeleyTrue** [comment]
> an additional hook(before) in the json that describes what unit tests to run for the challenge

@AngelKyriako Can you clarify this sentence? Perhaps with user stories.

**Comment by AngelKyriako** [comment]
@BerkeleyTrue Check this example: https://github.com/AngelKyriako/FccAdvancedNodejsExpressApp/blob/master/testApp/public/seeds.js. Its part of my custom test app which works if you clone it and run it yourself.

The value of the 'before' key will be evaluated and run, before the test cases start to run.

before, beforeEach, after & afterEach might be found helpful by other challenge unit tests in the future, but all I need is the before(all) one where I initialize my testing interface: https://github.com/AngelKyriako/FccAdvancedNodejsExpressApp/blob/master/testApp/public/challengeTester.js

**Comment by BerkeleyTrue** [comment]
@AngelKyriako Would this work in the tail section?

**Comment by AngelKyriako** [comment]
@BerkeleyTrue By 'tail section' u are talking about file formatting or runtime order? I need it to run before all the tests, I don't really mind where the hook will be located.

**Comment by BerkeleyTrue** [comment]
Challenges can have a head/tail property. This is code that is prepended/appended to the users code. This could potentially be used similar to a `before` in testing. It would run immediately after users code and before the tests run.

**Comment by AngelKyriako** [comment]
Cool, that should do the trick.

@BerkeleyTrue  Last time I checked I was lost while trying to import the challenges. I Could not find where to put the server's URL among other things. is there any documentation regarding the backend challenges setup. Is there a branch with the backend framewok in a stable state ?

**Comment by AngelKyriako** [comment]
@BerkeleyTrue @QuincyLarson 

I updated the readme with usage instruction & explanations on how to setup.

Let me know when you are ready, to port it together on the fcc site.

Cheers

**Comment by QuincyLarson** [comment]
@AngelKyriako awesome! Thanks for these additional updates and details. I'm working with @Em-Ant to get these into our codebase as quickly as possible :)

**Comment by AngelKyriako** [comment]
@QuincyLarson I was wondering what's going on regarding the challenges. Were they ever added in fcc ?

**Comment by AngelKyriako** [comment]
Looks great. From the high level overview I see, it seems quite better and more detailed that my initial design.

I will try to look it up in detail if I can put in the time.

Cheers & Good luck.

**Comment by QuincyLarson** [comment]
@AngelKyriako Awesome! I'm thrilled to hear you like what we've done with it. There's plenty of room for further expanding these and we'd welcome your help! :)
